src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -321,56 +320,45 @@ fromSnapLevels reg hfs hbio conf@TableConfig{..} uc resolve dir (SnapLevels leve
-        -- in supplyMergeCredits, then the resulting run is not tracked in the
-        -- registry, and closing the input runs is also not tracked in the
-        -- registry. Note, however, that this bit of code is likely to change in
-        -- #392.
-        let c = maybe 0 getUnspentCredits unspentCreditsMay
-              + maybe 0 getSpentCredits spentCreditsMay
-        supplyMergeCredits
-          (ScaledCredits c)
-          (creditThresholdForLevel conf ln)
-          incomingRun
-        pure Level{
-            residentRuns = snapResidentRuns
-          , ..
+        incomingRun <- fromSnapIncomingRun snapIncoming
+        pure Level {
+            incomingRun
+          , residentRuns = snapResidentRuns
           }
       where
         caching = diskCachePolicyForLevel confDiskCachePolicy ln
         alloc = bloomFilterAllocForLevel conf ln
 
         fromSnapIncomingRun ::
              SnapIncomingRun (Ref (Run m h))
-          -> m (Maybe UnspentCredits, Maybe SpentCredits, IncomingRun m h)
-        fromSnapIncomingRun (SnapMergingRun mpfl nr ne unspentCredits knownCompleted smrs) = do
-            (spentCreditsMay, mrs) <- fromSnapMergingRunState smrs
-            (Just unspentCredits, spentCreditsMay,) . Merging <$>
-              newMergingRun mpfl nr ne knownCompleted mrs
+          -> m (IncomingRun m h)
         fromSnapIncomingRun (SnapSingleRun run) =
-            pure (Nothing, Nothing, Single run)
-
-        fromSnapMergingRunState ::
-             SnapMergingRunState (Ref (Run m h))
-          -> m (Maybe SpentCredits, MergingRunState m h)
-        fromSnapMergingRunState (SnapCompletedMerge run) =
-            pure (Nothing, CompletedMerge run)
-        fromSnapMergingRunState (SnapOngoingMerge runs spentCredits mergeLast) = do
-            -- Initialse the variable with 0. Credits will be re-supplied later,
-            -- which will ensure that this variable is updated.
-            spentCreditsVar <- SpentCreditsVar <$> newPrimVar 0
-            rn <- uniqueToRunNumber <$> incrUniqCounter uc
-            mergeMaybe <- allocateMaybeTemp reg
-              (Merge.new hfs hbio caching alloc mergeLast resolve (mkPath rn) runs)
-              Merge.abort
-            case mergeMaybe of
-              Nothing -> error "openLevels: merges can not be empty"
-              Just m  -> pure (Just spentCredits, OngoingMerge runs spentCreditsVar m)
+            pure (Single run)
+        fromSnapIncomingRun (SnapMergingRun mpfl nr ne unspentCredits smrs) = do
+            Merging <$> case smrs of
+              SnapCompletedMerge run ->
+                MR.newCompleted mpfl nr ne run
+
+              SnapOngoingMerge runs spentCredits lvl -> do
+                rn <- uniqueToRunNumber <$> incrUniqCounter uc
+                mr <- allocateTemp reg
+                  (MR.new hfs hbio resolve caching alloc lvl mpfl (mkPath rn) runs)
+                  releaseRef
```

Comment:
True. It's slightly different, because here we don't roll back to a state where these runs existed, but we still run into a double free because they were created using `allocateTemp`, which already wants to clean them up.